### PR TITLE
[new release] mirage (2 packages) (4.5.0)

### PR DESCRIPTION
**Pull request by samoht**
The MirageOS library operating system

- Project page: <a href="https://github.com/mirage/mirage">https://github.com/mirage/mirage</a>
- Documentation: <a href="https://mirage.github.io/mirage/">https://mirage.github.io/mirage/</a>

##### CHANGES:

- This release introduces a significant change in the Mirage tool by
  splitting the definition of command-line arguments used at
  configure-time and runtime. Command-line arguments used in the
  configure script (also called 'configuration keys' and defined in
  the `Key` module) are essential during the setup of module
  dependencies for the unikernel, allowing for a specialized
  production of a unikernel for a given target runtime environment. On
  the other hand, command-line arguments that the unikernel can use at
  runtime (defined in the `Runtime_arg` module) are useful for
  customizing deployments without altering the dependencies of the
  unikernels. (mirage/mirage#1449, mirage/mirage#1450, mirage/mirage#1451, mirage/mirage#1455 @samoht, review by @hannesm)

  * API changes:
    - There is no more `~stage` parameter for `Key.Arg.info`.
    - `Key` now define command-line arguments for the configuration tool.
    - There is a new module `Runtime_arg` to define command-line arguments
      for the unikernel.
    - As there are no more keys type `'Both`, users are now expected to create
      two separated keys in that case (one for configure-time, one for runtime)
      or decide if the key is useful at runtime of configure-time.

  * Intended use of configuration keys (values of type `'a key`):
    - Used to set up module dependencies of the unikernel, such as the
      target (hvt, xen, etc.) and whether to use DHCP or a fixed IP address.
    - Enable the production of specialized unikernels suitable for
      specific target runtime environments and dedicated network and
      storage stacks.
    - Similar keys will produce reproducible binaries to be uploaded to artifact
      repositories like Docker Hub or https://builds.robur.coop/.

  * Intended use of command-line runtime arguments (values of type
    `a runtime_arg`):
    - Allow users to customize deployments by changing device
      configuration, like IP addresses, secrets, block device names,
      etc., post downloading of binaries.
    - These keys don’t alter the dependencies of the unikernels.
    - A runtime keys is just a reference to a normal Cmdliner term.

  * `key_gen.ml` is not generated anymore, so users cannot refer to
    `Key_gen.<key_name>` directy.
    - Any runtime argument has to be declared (using `runtime_arg` and
      registered on the device (using `~runtime_args`). The value of that
      argument will then be passed  as an extra parameter of the `connect`
      function of that device.
    - Configuration keys are not available at runtime anymore. For
      instance, `Key_gen.target` has been removed.

* Code migration:
    ```ocaml
    (* in config.ml *)
    let key =
      let doc = Key.Arg.info ~doc:"A Key." ~stage:`Run [ "key" ] in
      Key.(create "key" Arg.(opt_all ~stage:`Run string doc))
    ```
    becomes:
    ```ocaml
    (* in unikernel.ml *)
    open Cmdliner

    let key =
      let doc = Arg.info ~doc:"A Key." [ "key" ] in
      Arg.(value & opt_all string [] doc)
    ```

    ```ocaml
    (* in unikernel.ml *)
    let start _ =
      let key = Key_gen.hello () in
      ...
    ```
    becomes:
    ```ocaml
    (* in config.ml *)
    let hello = runtime_arg ~pos:__POS__ "Unikernel.hello"
    let main = main ~runtime_args:[hello] ...
    ```

    ```
    (* in unikernel.ml *)
    let hello =
      let open Cmdliner in
      let doc = Arg.info ~doc:"How to say hello." [ "hello" ] in
      Arg.(value & opt string "Hello World!" doc)

    let start _ hello =
      ...
    ```

- bump minimal ocaml-solo5 version bound to 0.8.2 to avoid fast memory usage
  error (mirage/mirage#1507, @palainp)
- BREAKING: the packages `functoria` and `functoria-runtime` are removed. The
  respectives libraries became `mirage.functoria` and `mirage-runtime.functoria`
  (mirage/mirage#1509, @samoht)
- BREAKING: `Mirage.keys` is renamed to `Mirage.runtime_args` (mirage/mirage#1506, @samoht)
- BREAKING: remove `Mirage.foreign. Use `Mirage.main` instead (mirage/mirage#1505, @samoht)
- BREAKING: `Mirage.main` does not take a `?extra_deps` parameter anymore
  (mirage/mirage#1505, @samoht)
- BREAKING: the `Mirage.connect` functions passed to create devices
  (with `Mirage.impl`)  now take values of type `'a Mirage.code` instead
  of `string`. Value of type `code` are created using a new `Mirage.code`
  function, that takes `~pos:__POS__` as parameter. This is used to generate
  better locations in the generated code, leading to better error messages
  (mirage/mirage#1504, @samoht)
- BUGFIX: fix `mirage describe` output (mirage/mirage#1446 @samoht), add test (mirage/mirage#1458 @samoht)
- Remove ipaddr from runtime (mirage/mirage#1437 @samoht, mirage/mirage#1465 @hannesm)
- BREAKING: `Mirage.register` no longer have `?packages` and `?keys` arguments
  (mirage/mirage#1433, mirage/mirage#1434 @hannesm)
- BREAKING: Remove deprecated bindings and types (mirage/mirage#1461 @hannesm)
- BREAKING: Remove `mirage build` subcommand, use `dune build` in Makefile
  (mirage/mirage#1404 @hannesm), put `--profile release` in Makefile instead of
  dune-workspace (mirage/mirage#1470 @hannesm)
- Increase dune version to 2.9 in generated dune-project (mirage/mirage#1443 @samoht)

### CODE REVIEW

File: packages/mirage/mirage.4.5.0/opam [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+depends: [
+  "ocaml" {>= "4.08.0"}
+  "dune" {>= "2.9.0"}
+  "astring"
+  "cmdliner" {>= "1.2.0"}
+  "emile" {>= "1.1"}
+  "fmt" {>= "0.8.7"}
+  "ipaddr" {>= "5.0.0"}
+  "bos"
```
Comment (samoht) [inline packages/mirage/mirage.4.5.0/opam:28]:
```suggestion
  "bos" {>= "0.7.0"}
```

**Comment by raphael-proust** [comment]
Some CI errors which would be better addressed:

1. `Result.bind`

```
#=== ERROR while compiling mirage.4.5.0 =======================================#
# context              2.2.0~beta2~dev | linux/x86_64 | ocaml-base-compiler.4.10.2 | pinned(https://github.com/mirage/mirage/releases/download/v4.5.0/mirage-4.5.0.tbz)
# path                 ~/.opam/4.10/.opam-switch/build/mirage.4.5.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p mirage -j 31
# exit-code            1
# env-file             ~/.opam/log/mirage-7-2cb08c.env
# output-file          ~/.opam/log/mirage-7-2cb08c.out
### output ###
# File "/home/opam/.opam/4.10/lib/bigstringaf/bigstringaf.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
# (cd _build/default && /home/opam/.opam/4.10/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/functoria/.functoria.objs/byte -I /home/opam/.opam/4.10/lib/angstrom -I /home/opam/.opam/4.10/lib/astring -I /home/opam/.opam/4.10/lib/base64 -I /home/opam/.opam/4.10/lib/bigstringaf -I /home/opam/.opam/4.10/lib/bos -I /home/opam/.opam/4.10/lib/cmdliner -I /home/opam/.opam/4.10/lib/domain-name -I /home/opam/.opam/4.10/lib/emile -I /home/opam/.opam/4.10/lib/fmt -I /home/opam/.opam/4.10/lib/fpath -I /home/opam/.opam/4.10/lib/ipaddr -I /home/opam/.opam/4.10/lib/logs -I /home/opam/.opam/4.10/lib/macaddr -I /home/opam/.opam/4.10/lib/pecu -I /home/opam/.opam/4.10/lib/result -I /home/opam/.opam/4.10/lib/rresult -I /home/opam/.opam/4.10/lib/stdlib-shims -I /home/opam/.opam/4.10/lib/stringext -I /home/opam/.opam/4.10/lib/uchar -I /home/opam/.opam/4.10/lib/uri -I /home/opam/.opam/4.10/lib/uutf -intf-suffix .ml -no-alias-deps -open Functoria__ -o lib/functoria/.functoria.objs/byte/functoria__Tool.cmo -c -impl lib/functoria/tool.ml)
# File "lib/functoria/tool.ml", line 34, characters 17-28:
# 34 |   let ( let* ) = Result.bind in
#                       ^^^^^^^^^^^
# Error: Unbound value Result.bind
# (cd _build/default && /home/opam/.opam/4.10/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/functoria/.functoria.objs/byte -I /home/opam/.opam/4.10/lib/angstrom -I /home/opam/.opam/4.10/lib/astring -I /home/opam/.opam/4.10/lib/base64 -I /home/opam/.opam/4.10/lib/bigstringaf -I /home/opam/.opam/4.10/lib/bos -I /home/opam/.opam/4.10/lib/cmdliner -I /home/opam/.opam/4.10/lib/domain-name -I /home/opam/.opam/4.10/lib/emile -I /home/opam/.opam/4.10/lib/fmt -I /home/opam/.opam/4.10/lib/fpath -I /home/opam/.opam/4.10/lib/ipaddr -I /home/opam/.opam/4.10/lib/logs -I /home/opam/.opam/4.10/lib/macaddr -I /home/opam/.opam/4.10/lib/pecu -I /home/opam/.opam/4.10/lib/result -I /home/opam/.opam/4.10/lib/rresult -I /home/opam/.opam/4.10/lib/stdlib-shims -I /home/opam/.opam/4.10/lib/stringext -I /home/opam/.opam/4.10/lib/uchar -I /home/opam/.opam/4.10/lib/uri -I /home/opam/.opam/4.10/lib/uutf -intf-suffix .ml -no-alias-deps -open Functoria__ -o lib/functoria/.functoria.objs/byte/functoria__Lib.cmo -c -impl lib/functoria/lib.ml)
# File "lib/functoria/lib.ml", line 287, characters 43-56:
# 287 |         let cwd = Bos.OS.Dir.current () |> Result.get_ok in
#                                                  ^^^^^^^^^^^^^
# Error: Unbound value Result.get_ok
# (cd _build/default && /home/opam/.opam/4.10/bin/ocamlopt.opt -w -40 -g -I lib/functoria/.functoria.objs/byte -I lib/functoria/.functoria.objs/native -I /home/opam/.opam/4.10/lib/angstrom -I /home/opam/.opam/4.10/lib/astring -I /home/opam/.opam/4.10/lib/base64 -I /home/opam/.opam/4.10/lib/bigstringaf -I /home/opam/.opam/4.10/lib/bos -I /home/opam/.opam/4.10/lib/cmdliner -I /home/opam/.opam/4.10/lib/domain-name -I /home/opam/.opam/4.10/lib/emile -I /home/opam/.opam/4.10/lib/fmt -I /home/opam/.opam/4.10/lib/fpath -I /home/opam/.opam/4.10/lib/ipaddr -I /home/opam/.opam/4.10/lib/logs -I /home/opam/.opam/4.10/lib/macaddr -I /home/opam/.opam/4.10/lib/pecu -I /home/opam/.opam/4.10/lib/result -I /home/opam/.opam/4.10/lib/rresult -I /home/opam/.opam/4.10/lib/stdlib-shims -I /home/opam/.opam/4.10/lib/stringext -I /home/opam/.opam/4.10/lib/uchar -I /home/opam/.opam/4.10/lib/uri -I /home/opam/.opam/4.10/lib/uutf -intf-suffix .ml -no-alias-deps -open Functoria__ -o lib/functoria/.functoria.objs/native/functoria__Tool.cmx -c -impl lib/functoria/tool.ml)
# File "lib/functoria/tool.ml", line 34, characters 17-28:
# 34 |   let ( let* ) = Result.bind in
#                       ^^^^^^^^^^^
# Error: Unbound value Result.bind
# (cd _build/default && /home/opam/.opam/4.10/bin/ocamlopt.opt -w -40 -g -I lib/functoria/.functoria.objs/byte -I lib/functoria/.functoria.objs/native -I /home/opam/.opam/4.10/lib/angstrom -I /home/opam/.opam/4.10/lib/astring -I /home/opam/.opam/4.10/lib/base64 -I /home/opam/.opam/4.10/lib/bigstringaf -I /home/opam/.opam/4.10/lib/bos -I /home/opam/.opam/4.10/lib/cmdliner -I /home/opam/.opam/4.10/lib/domain-name -I /home/opam/.opam/4.10/lib/emile -I /home/opam/.opam/4.10/lib/fmt -I /home/opam/.opam/4.10/lib/fpath -I /home/opam/.opam/4.10/lib/ipaddr -I /home/opam/.opam/4.10/lib/logs -I /home/opam/.opam/4.10/lib/macaddr -I /home/opam/.opam/4.10/lib/pecu -I /home/opam/.opam/4.10/lib/result -I /home/opam/.opam/4.10/lib/rresult -I /home/opam/.opam/4.10/lib/stdlib-shims -I /home/opam/.opam/4.10/lib/stringext -I /home/opam/.opam/4.10/lib/uchar -I /home/opam/.opam/4.10/lib/uri -I /home/opam/.opam/4.10/lib/uutf -intf-suffix .ml -no-alias-deps -open Functoria__ -o lib/functoria/.functoria.objs/native/functoria__Lib.cmx -c -impl lib/functoria/lib.ml)
# File "lib/functoria/lib.ml", line 287, characters 43-56:
# 287 |         let cwd = Bos.OS.Dir.current () |> Result.get_ok in
#                                                  ^^^^^^^^^^^^^
# Error: Unbound value Result.get_ok
```

This is likely because of a transitive dependency to `result` or some `shims` library which gets downgraded with the other downgrades? The error occurs on ocaml<4.14

2. package does not exist

```
#=== ERROR while compiling mirage.4.5.0 =======================================#
# context              2.2.0~beta2~dev | linux/x86_64 | ocaml-base-compiler.4.10.2 | pinned(https://github.com/mirage/mirage/releases/download/v4.5.0/mirage-4.5.0.tbz)
# path                 ~/.opam/4.10/.opam-switch/build/mirage.4.5.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune runtest -p mirage -j 31
# exit-code            1
# env-file             ~/.opam/log/mirage-7-0f7da6.env
# output-file          ~/.opam/log/mirage-7-0f7da6.out
### output ###
# File "test/functoria/e2e/dune", line 20, characters 11-25:
# 20 |   (package mirage-runtime))
#                 ^^^^^^^^^^^^^^
# Error: Package mirage-runtime does not exist
# (cd _build/default/test/mirage && ./test.exe)
# [THE REST OF THE LOGS SHOW SUCCESSFUL TESTS]
```

This is likely due to a packaging issue where the opam file for `mirage-runtime` is not included in the tarball for `mirage` but the package is mentioned in the `dune` files anyway. We can leave with this false negative test but it would be better to fix it if possible (split the tests for different packages in different directories and tarball just what's needed? Maybe too complex Idk what the repo looks like.


----------------


All other errors seem duplicate of those or unrelated revdeps failures

File: packages/mirage/mirage.4.5.0/opam (hunk 1 above)
Comment (raphael-proust) [reply packages/mirage/mirage.4.5.0/opam]: That's probably the fix for the `Result.bind` issue mentioned in the comment below.

File: packages/mirage/mirage.4.5.0/opam [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  "emile" {>= "1.1"}
+  "fmt" {>= "0.8.7"}
+  "ipaddr" {>= "5.0.0"}
+  "bos" {>= "0.7.0"}
+  "fpath"
+  "rresult" {>= "0.7.0"}
+  "uri" {>= "4.2.0"}
+  "logs"
+  "opam-monorepo" {>= "0.3.2"}
+  "alcotest" {with-test}
```
Comment (samoht) [inline packages/mirage/mirage.4.5.0/opam:34]:
```suggestion
  "alcotest" {with-test}
  "mirage-runtime" {with-test & = version}
```

**Comment by samoht** [comment]
Thanks, I've tried to address both issues with additional commits/suggestions

**Comment by mseri** [comment]
Thanks

**Comment by samoht** [comment]
Hum actually I might have to revert this, the release is broken because of https://github.com/mirage/mirage/pull/1518

**Comment by mseri** [comment]
CAn you release a 4.5.1 and mark it unavailable? Otherwise, since it has literally been just released, for such a small change you could add a patch to this

**Pull request merged**